File: linkList/7.removeNthNodeFromEnd/go/removeNthNodeFromEnd.go
package _go
type ListNode struct {
	Val int
	Next *ListNode
	}
	func removeNthFromEnd(head *ListNode, n int) *ListNode {
		if head == nil {
			return nil
		}
		before,after := head,head
		// 前面指针先走n步
		i:=0
		for ;before != nil && i < n;i++{
			before = before.Next
		}
		if before == nil {
			if i < n {
				// 总长度不够n,返回原值
				return head
			}else {
				// 链表长度恰好为n,删除第一个节点
				return head.Next
			}
		}
		// 后面指针的前一个值
		var pre *ListNode
		// 两个指针同时走，前面的走到末尾，后面的正好到要删除的值上
		for before != nil {
			pre = after
			after = after.Next
			before = before.Next
		}
		// 删除指定位置节点
		pre.Next = after.Next
		return head
	}

File: linkList/7.removeNthNodeFromEnd/py/removeNthNodeFromEnd.py
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if not head:
            return None
        # 定义前后两个指针
        before, after = head, head
        # 前指针先走n步
        i = 0
        while before and i < n:
            before = before.next
            i += 1
        if not before:
            if i < n:
                # 长度不足n,原样返回
                return head
            else:
                # 长度恰好为n,删除头节点
                return head.next
        # 指向前一个节点，用于删除节点
        pre: Optional[ListNode] = None
        # 前后指针一起走，前面走到末尾，后面正好到倒数第n个
        while before:
            pre = after
            after = after.next
            before = before.next
        pre.next = after.next
        return head